paml/paml.ttl

```diff
@@ -541,3 +541,151 @@ paml:fromSamples rdf:type owl:ObjectProperty ;
               rdfs:domain paml:LocatedData ;
               rdfs:range sbol:LocatedSamples ;
               rdfs:label "from_samples" .
+
+#################################################################
+#    Time properties
+#################################################################
+
+# <timeproperty> ::= prov:startedAtTime | prov:endedAtTime
+# <timevariable> ::= <URI for Protocol, Activity, ...> valueOf <timeproperty>
+# <duration> ::= Difference(<timevariable>, <timevariable>) | OM:Measure
+# <expression> ::= <duration>| <float> | <float> * <expression> |  <expression> - <expression> | <expression> + <expression>
+# <bincomp> ::= <= | < | >= | > | =
+# <binlog> ::= AND | OR | XOR
+# <constraint> ::= <expression> <bincomp> <expression> | <constraint> <binlog> <constraint> | NOT <constraint>
+
+paml:timeVariableValue rdf:type owl:ObjectProperty ;
+                rdfs:subPropertyOf opil:compositionalProperty ;
+                rdfs:comment "Time value" ;
+                rdfs:label "value" ;
+                rdfs:domain paml:TimeVariable ;
+                rdfs:range sbol:Measure .
+
+paml:TimeVariable rdf:type owl:Class ;
```

Comment:
This doesn't seem to be tracking the sketch of timevariable above. Shouldn't timevariable be pointing to a protocol rather than a measure, and the measure be on the duration?

Reply:
I realize it deviates a bit. I think it’s as it should be in the current form. We could also link back the timevariable to its parent if that is what you mean. Currently this is liking from activity to to variable and variable to value/measure.

Reply:
I'm not sure that I agree. In this form, it seems to be going back to a notion of absolute time, rather than time relationships, and I don't see where the constraints are being stored.

Reply:
I see.  There are two main ways to represent time information about activities (in the general sense, not necessarily PAML).  Either with points or intervals.  It turns out that constraints between metric points allow you to express both.  

I did this first version without explicit constraints because I wanted to get the timepoints expressed.  Our compilation into constraints uses the graph structure of the protocol to express the implicit constraints.  Explicit constraints is the next step.

Reply:
Right: and what we'd captured in the model on lines 549-555 captures this as a network of measurement constraints that points to the items they describe and their relations.

What you've got right now is describing the relationships on a specific execution of a protocol, rather than on a protocol as a program that can be executed at multiple different times.

Reply:
Let me summarize the issues here:

- The current representation maps from protocol elements to time variables.  The time variables *possibly* map to values.  My thinking was that the time variables that are bound will correspond to values chosen while treating the protocol as a program.  Calling paml-check will bind additional unbound variables to represent a possible execution.  We could have also bound these variables during execution.  I think that we should possibly distinguish between given values and computed/observed values.  

- You were expecting an explicit constraint network that links variables back to their protocol elements.  I think this could get cumbersome because many of the constraints are implied by the protocol structure.  We are currently computing these from the protocol when we get to paml-check.  We could, however, compute and represent them in PAML.  I don't see a particular advantage to this, but its possible.  

- We need to add explicit constraints that are not implied by the protocol structure in the way you were expecting (previous point).    These would go along with the protocol to be checked by paml-check.  

So, there are a few questions about where do we go from here:

1. What should paml-check return?  We currently compute a possible execution and fill in missing time values in the protocol.  I think that maybe we should return yes/no "is this protocol consistent?"  This will avoid the issue about computed values.  We can create other types of checks in the future, such as what we currently have.  A possible next call would compute the minimum duration of the protocol.
2. Should we make the implicit protocol constraints explicit in the RDF?  I think this is a no.
3. Should we include explicit constraints in this PR?  I say no, but am flexible.

Reply:
I believe that it's significant that in your tests, that only thing that's getting assigned is either "time = 0" (which is semantically incorrect, per note below, since time is a quantity rather than a unit) or non-zero durations.

I do think that it's appropriate to keep most of the relations implicit, and to just extract them for the solver. But I think that what we need to be representing here is, in fact, the explicit constraints that cannot be inferred. We need a place to say "executing primitive "incubate" with parameter X equal to duration D implies that the primitive will have execution duration of at least D" and "the time between protocol start and end of step 7 must be less than 6 hours."

We can then run checks in two different modes:
1. Check the durations of a protocol to see if it is impossible to execute correctly
2. Check an execution trace (live or retrospectively) to see if anything was violated

In both cases, we should *not* ever be marking up the actual protocol with actual time points --- we should be doing it the other way around, with a collection of time points in our inference that refer to points in the protocol.